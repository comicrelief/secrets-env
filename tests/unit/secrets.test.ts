import SecretsManager from 'aws-sdk/clients/secretsmanager';
import sinon from 'sinon';

import {
  getSecret,
  getSecrets,
  secretsJson,
  secretsBash,
} from '../../src/secrets';
import {
  listFormat,
  valueFormat,
  SECRETS,
  SECRETS_LIST,
} from '../mocks/secrets.mock';

const getContext = (
  single: any = SECRETS.SECRETS_ONE_NAMESPACE_TWO_STAGES,
  list: any = SECRETS_LIST,
  namespace = 'your-service-1',
  stage = 'staging'
) => {
  const client = {
    getSecretValue: sinon.fake.returns({ promise: sinon.fake.resolves(single) }),
    listSecrets: sinon.fake.returns({ promise: sinon.fake.resolves(list) }),
  } as unknown as SecretsManager;

  const context = {
    client,
    command: 'json',
    describe: true,
    namespace,
    stage,
    verbose: false,
  };

  return context;
}

describe('unit.secrets', () => {

  afterEach(() => sinon.restore());

  describe('getSecret', () => {
    describe('Returns the JSON formatted secret', () => {
      Object.entries(SECRETS).forEach(([key, secret]) => {
        it(`Secret: ${key}`, async () => {
          const input = listFormat(secret);
          const metadata = valueFormat(secret);
          const context = getContext(metadata)

          const actual = await getSecret(context, input);

          expect(actual).toMatchSnapshot();
        })
      })
    })
  })

  describe('getSecrets', () => {
    it('Returns the JSON formatted secrets', async () => {
      const context = getContext(undefined, SECRETS_LIST);

      const actual = await getSecrets(context);

      expect(actual).toMatchSnapshot();
    })
  })

  describe('secretsJson', () => {
    it('Returns the JSON formatted secrets', async () => {
      const context = getContext(undefined, SECRETS_LIST);

      const actual = await secretsJson(context);

      expect(actual).toMatchSnapshot();
    })
  })

  describe('secretsBash', () => {
    it('Returns the Bash formatted secrets', async () => {
      const context = getContext(undefined, SECRETS_LIST);

      const actual = await secretsBash(context);

      expect(actual).toMatchSnapshot();
    })
  })

});
