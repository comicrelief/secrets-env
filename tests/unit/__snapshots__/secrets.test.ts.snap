// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit.secrets getSecret Returns the JSON formatted secret Secret: SECRETS_NO_TAG 1`] = `
Object {
  "ARN": "arn:aws:secretsmanager:eu-west-1:some-account858584:secret:SECRET_NO_TAG",
  "Description": "",
  "Name": "SECRET_NO_TAG",
  "Path": "staging/your-namespace-1/SECRET_NO_TAG",
  "Tags": Array [
    Object {
      "Key": "SomeTag",
      "Value": "SomeValue",
    },
  ],
  "Value": "SECRET-SECRETS_NO_TAG",
}
`;

exports[`unit.secrets getSecret Returns the JSON formatted secret Secret: SECRETS_ONE_NAMESPACE_TWO_STAGES 1`] = `
Object {
  "ARN": "arn:aws:secretsmanager:eu-west-1:some-account858584:secret:SECRETS_ONE_NAMESPACE_TWO_STAGES",
  "Description": "A key used in one namespace only, but in two stage environments",
  "Name": "SECRETS_ONE_NAMESPACE_TWO_STAGES",
  "Path": "staging/your-namespace-1/SECRETS_ONE_NAMESPACE_TWO_STAGES",
  "Tags": Array [
    Object {
      "Key": "env:your-service-1/dev",
      "Value": "",
    },
    Object {
      "Key": "env:your-service-1/staging",
      "Value": "",
    },
  ],
  "Value": "SECRET-SECRETS_ONE_NAMESPACE_TWO_STAGES",
}
`;

exports[`unit.secrets getSecret Returns the JSON formatted secret Secret: SECRETS_TWO_NAMESPACES_THREE_STAGES 1`] = `
Object {
  "ARN": "arn:aws:secretsmanager:eu-west-1:some-account858584:secret:SECRETS_TWO_NAMESPACES_THREE_STAGES",
  "Description": "A key used in two namespaces for a total of three stages",
  "Name": "SECRETS_TWO_NAMESPACES_THREE_STAGES",
  "Path": "staging/your-namespace-2/SECRETS_TWO_NAMESPACES_THREE_STAGES",
  "Tags": Array [
    Object {
      "Key": "env:your-service-2/staging",
      "Value": "",
    },
    Object {
      "Key": "env:your-service-1/dev",
      "Value": "",
    },
    Object {
      "Key": "env:your-service-1/staging",
      "Value": "",
    },
  ],
  "Value": "SECRET-SECRETS_TWO_NAMESPACES_THREE_STAGES",
}
`;

exports[`unit.secrets getSecrets Returns the JSON formatted secrets 1`] = `
Array [
  Object {
    "ARN": "arn:aws:secretsmanager:eu-west-1:some-account858584:secret:SECRET_NO_TAG",
    "Description": "",
    "Name": "SECRET_NO_TAG",
    "Path": "staging/your-namespace-1/SECRET_NO_TAG",
    "Tags": Array [
      Object {
        "Key": "SomeTag",
        "Value": "SomeValue",
      },
    ],
    "Value": "SECRET-SECRETS_ONE_NAMESPACE_TWO_STAGES",
  },
  Object {
    "ARN": "arn:aws:secretsmanager:eu-west-1:some-account858584:secret:SECRETS_ONE_NAMESPACE_TWO_STAGES",
    "Description": "A key used in one namespace only, but in two stage environments",
    "Name": "SECRETS_ONE_NAMESPACE_TWO_STAGES",
    "Path": "staging/your-namespace-1/SECRETS_ONE_NAMESPACE_TWO_STAGES",
    "Tags": Array [
      Object {
        "Key": "env:your-service-1/dev",
        "Value": "",
      },
      Object {
        "Key": "env:your-service-1/staging",
        "Value": "",
      },
    ],
    "Value": "SECRET-SECRETS_ONE_NAMESPACE_TWO_STAGES",
  },
  Object {
    "ARN": "arn:aws:secretsmanager:eu-west-1:some-account858584:secret:SECRETS_TWO_NAMESPACES_THREE_STAGES",
    "Description": "A key used in two namespaces for a total of three stages",
    "Name": "SECRETS_TWO_NAMESPACES_THREE_STAGES",
    "Path": "staging/your-namespace-2/SECRETS_TWO_NAMESPACES_THREE_STAGES",
    "Tags": Array [
      Object {
        "Key": "env:your-service-2/staging",
        "Value": "",
      },
      Object {
        "Key": "env:your-service-1/dev",
        "Value": "",
      },
      Object {
        "Key": "env:your-service-1/staging",
        "Value": "",
      },
    ],
    "Value": "SECRET-SECRETS_ONE_NAMESPACE_TWO_STAGES",
  },
]
`;

exports[`unit.secrets secretsBash Returns the Bash formatted secrets 1`] = `
"export SECRET_NO_TAG=\\"SECRET-SECRETS_ONE_NAMESPACE_TWO_STAGES\\"                       #  [your-service-1, staging]
export SECRETS_ONE_NAMESPACE_TWO_STAGES=\\"SECRET-SECRETS_ONE_NAMESPACE_TWO_STAGES\\"    # A key used in one namespace only, but in two stage environments [your-service-1, staging]
export SECRETS_TWO_NAMESPACES_THREE_STAGES=\\"SECRET-SECRETS_ONE_NAMESPACE_TWO_STAGES\\" # A key used in two namespaces for a total of three stages [your-service-1, staging]"
`;

exports[`unit.secrets secretsJson Returns the JSON formatted secrets 1`] = `
Array [
  Object {
    "ARN": "arn:aws:secretsmanager:eu-west-1:some-account858584:secret:SECRET_NO_TAG",
    "Description": "",
    "Name": "SECRET_NO_TAG",
    "Path": "staging/your-namespace-1/SECRET_NO_TAG",
    "Tags": Array [
      Object {
        "Key": "SomeTag",
        "Value": "SomeValue",
      },
    ],
    "Value": "SECRET-SECRETS_ONE_NAMESPACE_TWO_STAGES",
  },
  Object {
    "ARN": "arn:aws:secretsmanager:eu-west-1:some-account858584:secret:SECRETS_ONE_NAMESPACE_TWO_STAGES",
    "Description": "A key used in one namespace only, but in two stage environments",
    "Name": "SECRETS_ONE_NAMESPACE_TWO_STAGES",
    "Path": "staging/your-namespace-1/SECRETS_ONE_NAMESPACE_TWO_STAGES",
    "Tags": Array [
      Object {
        "Key": "env:your-service-1/dev",
        "Value": "",
      },
      Object {
        "Key": "env:your-service-1/staging",
        "Value": "",
      },
    ],
    "Value": "SECRET-SECRETS_ONE_NAMESPACE_TWO_STAGES",
  },
  Object {
    "ARN": "arn:aws:secretsmanager:eu-west-1:some-account858584:secret:SECRETS_TWO_NAMESPACES_THREE_STAGES",
    "Description": "A key used in two namespaces for a total of three stages",
    "Name": "SECRETS_TWO_NAMESPACES_THREE_STAGES",
    "Path": "staging/your-namespace-2/SECRETS_TWO_NAMESPACES_THREE_STAGES",
    "Tags": Array [
      Object {
        "Key": "env:your-service-2/staging",
        "Value": "",
      },
      Object {
        "Key": "env:your-service-1/dev",
        "Value": "",
      },
      Object {
        "Key": "env:your-service-1/staging",
        "Value": "",
      },
    ],
    "Value": "SECRET-SECRETS_ONE_NAMESPACE_TWO_STAGES",
  },
]
`;
